// oauth-v2-client/dist/grant-controls/AuthorizationCodePKCEGrantControl.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var common_1 = require("@noreajs/common");
var query_string_1 = require("query-string");
var helpers_1 = require("../helpers");
var basicAuthFunc_1 = __importDefault(require("../helpers/basicAuthFunc"));
var pkceFactory_1 = require("../helpers/pkceFactory");
var GrantControl_1 = __importDefault(require("./GrantControl"));
var AuthorizationCodePKCEGrantControl = /** @class */ (function (_super) {
    __extends(AuthorizationCodePKCEGrantControl, _super);
    function AuthorizationCodePKCEGrantControl(config, options) {
        var _this = _super.call(this, config) || this;
        _this.options = options;
        // callback url
        _this.redirectUri = _this.options.callbackUrl;
        return _this;
    }
    /**
     * Get authentication url
     * @param {GetAuthorizationUrlFuncConfig} options redirect uri, response type
     */
    AuthorizationCodePKCEGrantControl.prototype.getAuthUri = function (options) {
        var _a, _b, _c, _d, _e;
        // create local properties
        var localState = options === null || options === void 0 ? void 0 : options.state;
        var localCodeVerifier = options === null || options === void 0 ? void 0 : options.codeVerifier;
        var localCodeChallengeMethod = (_a = options === null || options === void 0 ? void 0 : options.codeChallengeMethod) !== null && _a !== void 0 ? _a : this.options.codeChallengeMethod;
        var localCodeChallenge = localCodeVerifier ? localCodeChallengeMethod === "S256" ? (0, pkceFactory_1.generateCodeChallenge)(localCodeVerifier) : localCodeVerifier : undefined;
        var localScopes = (_b = options === null || options === void 0 ? void 0 : options.scopes) !== null && _b !== void 0 ? _b : this.options.scopes;
        var localRedirectUrl = (_c = options === null || options === void 0 ? void 0 : options.callbackUrl) !== null && _c !== void 0 ? _c : this.redirectUri;
        // query params
        var queryParams = {
            response_type: (_d = options === null || options === void 0 ? void 0 : options.responseType) !== null && _d !== void 0 ? _d : "code",
            redirect_uri: localRedirectUrl,
            client_id: this.options.clientId,
            state: localState,
            scope: localScopes ? localScopes.join(" ") : "",
            code_challenge: localCodeChallenge,
            code_challenge_method: localCodeChallengeMethod,
        };
        // merged params
        var mergedParams = common_1.Obj.cleanWithEmpty(common_1.Obj.merge(queryParams, (_e = this.requestOptions.query) !== null && _e !== void 0 ? _e : {}));
        // constructing the request
        var url = new URL(this.options.authUrl);
        for (var param in mergedParams) {
            if (Object.prototype.hasOwnProperty.call(mergedParams, param)) {
                var value = mergedParams[param];
                // setting the param
                url.searchParams.set(param, value);
            }
        }
        return url.toString();
    };
    /**
     * Get token with the authorization code extracted in the callback uri
     * @param params {GetAuthorizationTokenFuncConfig} parameters
     */
    AuthorizationCodePKCEGrantControl.prototype.getToken = function (params) {
        var _a, _b;
        return __awaiter(this, void 0, void 0, function () {
            var urlData_1, localState, localCodeVerifier, localRedirectUrl, requestHeaders, requestBody, error_1;
            return __generator(this, function (_c) {
                switch (_c.label) {
                    case 0:
                        _c.trys.push([0, 4, , 5]);
                        urlData_1 = (0, query_string_1.parseUrl)(params.callbackUrl);
                        localState = params.state;
                        localCodeVerifier = params.codeVerifier;
                        localRedirectUrl = (_a = params === null || params === void 0 ? void 0 : params.redirectUri) !== null && _a !== void 0 ? _a : this.redirectUri;
                        // force array
                        if (localState === null || localState === undefined) {
                            localState = [];
                        }
                        else if (!Array.isArray(localState)) {
                            localState = [localState];
                        }
                        // state exists
                        if (urlData_1.query.state && !localState.includes(urlData_1.query.state)) {
                            if (this.log === true || params.log === true) {
                                console.log("Corrupted answer, the state doesn't match.", {
                                    urlData: urlData_1,
                                    localState: localState,
                                });
                            }
                            throw new Error("Corrupted answer, the state doesn't match.");
                        }
                        if (!urlData_1.query.code) return [3 /*break*/, 2];
                        requestHeaders = {};
                        requestBody = {
                            grant_type: "authorization_code",
                            code: urlData_1.query.code,
                            state: localState,
                            redirect_uri: localRedirectUrl,
                            code_verifier: localCodeVerifier,
                        };
                        /**
                         * Client authentication
                         * ----------------------
                         */
                        if (this.options.basicAuthHeader === false) {
                            requestBody["client_id"] = this.options.clientId;
                            requestBody["client_secret"] = this.options.clientSecret;
                        }
                        else {
                            requestHeaders["Authorization"] = (0, basicAuthFunc_1.default)(this.options.clientId, (_b = this.options.clientSecret) !== null && _b !== void 0 ? _b : "");
                        }
                        /**
                         * Request a token
                         */
                        return [4 /*yield*/, (0, helpers_1.requestToken)({
                                accessTokenUrl: this.options.accessTokenUrl,
                                body: requestBody,
                                config: {
                                    oauthOptions: this.oauthOptions,
                                    requestOptions: this.requestOptions,
                                },
                                headers: requestHeaders,
                                onError: params.onError,
                                onSuccess: function (data) {
                                    // call the parent token
                                    if (params.onSuccess)
                                        params.onSuccess(data, urlData_1.query.state);
                                },
                                requestOptions: params.requestOptions
                            })];
                    case 1:
                        /**
                         * Request a token
                         */
                        _c.sent();
                        return [3 /*break*/, 3];
                    case 2: throw urlData_1.query;
                    case 3: return [3 /*break*/, 5];
                    case 4:
                        error_1 = _c.sent();
                        if (params.onError) {
                            params.onError(error_1);
                        }
                        else {
                            throw error_1;
                        }
                        return [3 /*break*/, 5];
                    case 5: return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Refresh the token
     * @param params parameters
     */
    AuthorizationCodePKCEGrantControl.prototype.refresh = function (params) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, (0, helpers_1.refreshToken)({
                            accessTokenUrl: this.options.accessTokenUrl,
                            config: {
                                oauthOptions: this.oauthOptions,
                                requestOptions: this.requestOptions,
                            },
                            onSuccess: function (data) {
                                // this update token
                                if (params.onSuccess)
                                    params.onSuccess(data);
                            },
                            params: params
                        })];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    return AuthorizationCodePKCEGrantControl;
}(GrantControl_1.default));
exports.default = AuthorizationCodePKCEGrantControl;
